src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -1989,30 +1993,86 @@ GenerateGrantOnSchemaQueriesFromAclItem(Oid schemaOid, AclItem *aclItem)
 }
 
 
-GrantStmt *
-GenerateGrantOnSchemaStmtForRights(Oid roleOid,
-								   Oid schemaOid,
-								   char *permission,
-								   bool withGrantOption)
+/*
+ * GenerateGrantStmtForRights is the function for creating GrantStmt's for all
+ * types of objects that are supported. It takes parameters to fill a GrantStmt's
+ * fields and returns the GrantStmt.
+ * The field `objects` of GrantStmt doesn't have a common structure for all types.
+ * Make sure you have added your object type to GetObjectsForGrantStmt.
+ */
+static GrantStmt *
+GenerateGrantStmtForRights(ObjectType objectType,
+						   Oid roleOid,
+						   Oid objectId,
+						   char *permission,
+						   bool withGrantOption)
+{
+	GrantStmt *stmt = makeNode(GrantStmt);
+	stmt->is_grant = true;
+	stmt->targtype = ACL_TARGET_OBJECT;
+	stmt->objtype = objectType;
+	stmt->objects = GetObjectsForGrantStmt(objectType, objectId);
+	stmt->privileges = list_make1(GetAccessPrivObjectForGrantStmt(permission));
+	stmt->grantees = list_make1(GetRoleSpecObjectForGrantStmt(roleOid));
+	stmt->grant_option = withGrantOption;
+
+	return stmt;
+}
+
+
+/*
+ * GetObjectsForGrantStmt takes an object type and object id and returns the 'objects'
+ * field to be used when creating GrantStmt.
+ */
```

Comment:
Also add a comment here to show that we always return a list with 1 item but we need to return list because it can contain different types of data on enterprise.